pkg/auth/providers/saml/saml_provider_test.go

```diff
@@ -0,0 +1,147 @@
+package saml
+
+import (
+	"context"
+	"github.com/rancher/norman/types"
+	"github.com/rancher/rancher/pkg/auth/providers/ldap"
+	"github.com/rancher/rancher/pkg/auth/tokens"
+	v3 "github.com/rancher/rancher/pkg/generated/norman/management.cattle.io/v3"
+	"github.com/rancher/rancher/pkg/types/config"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/rest"
+	"testing"
+)
+
+func createMockProvider(t *testing.T, name string, hasLdap bool) *Provider {
+	// saml.Configure runs some ldap specific logic based on the saml provider name, so we provide
+	// just enough scaffolding to create the provider that way.
+	ctx := context.Background()
+	mgmtCtx, err := config.NewScaledContext(rest.Config{}, nil)
+	if err != nil {
+		t.Fatal("Failed to create NewScaledContext")
+	}
+	tokenMGR := tokens.NewManager(ctx, mgmtCtx)
+	configuredProvider, ok := Configure(ctx, mgmtCtx, mgmtCtx.UserManager, tokenMGR, "okta").(*Provider)
+	if !ok {
+		t.Fatal("Failed to Configure a valid Provider")
+	}
+	return configuredProvider
+}
+
+func TestConfiguredOktaProviderContainsLdapProvider(t *testing.T) {
+	provider := createMockProvider(t, "okta", false)
+
+	if !provider.hasLdapGroupSearch() {
+		t.Error("Missing LDAP group search capability for okta provider")
+	}
+
+	if provider.ldapProvider == nil {
+		t.Error("Configured okta provider did not receive child LDAP provider")
+	}
+}
+
+func TestProvidedLdapUsedForPrincipalSearch(t *testing.T) {
+	provider := createMockProvider(t, "okta", false)
+	provider.ldapProvider = &mockLdapProvider{isLdapConfigured: true, providerName: provider.name}
+
+	// Our mock provider has just a single user, "alice". If it was called, we'll have that
+	// user's full name, not a dummy from the incomplete search key.
+	results, err := provider.SearchPrincipals("al", "user", v3.Token{})
+	if err != nil {
+		t.Fatalf("Failed to search principals: %s", err)
+	}
+	if results[0].LoginName != "alice" {
+		t.Errorf("Expected LoginName 'alice', got '%s'", results[0].LoginName)
+	}
+}
+
+func TestFallbackBehaviorForNonLdapPrincipalSearch(t *testing.T) {
+	provider := createMockProvider(t, "okta", false)
+	provider.ldapProvider = &mockLdapProvider{isLdapConfigured: false, providerName: provider.name}
+
+	results, err := provider.SearchPrincipals("al", "user", v3.Token{})
+	if err != nil {
+		t.Fatalf("Failed to search principals: %s", err)
+	}
+	// Since our mock provider does not have ldap capabilities, the principal we receive here should
+	// come from saml, and match our searchKey exactly.
+	if results[0].LoginName != "al" {
+		t.Errorf("Expected generated LoginName 'al', got '%s'", results[0].LoginName)
+	}
+}
```

Comment:
Consider combining these tests (and possibly others) as a table driven test. For example:
```
testcases := []struct{
  name              string
  expectedLoginName string
  expectError       bool
} {
    {
        name: "test fallback behaviour",
        expectedLoginName: "al",
        expectError: false,
    },
    {
        name: "test provided ldap user for login",
        expectedLoginName: "alice",
        expectError: false,
    },
}

for _, tt := range testcases {
    t.Run(t, tt.name, func(t *testing.T) {
        results, err := provider.SearchPrincipals(...)
        
        if tt.expectError {
            assert.Error(t, err)
            return
        }
        
        require.NoError(t, err)
        ...
        assert.Equal(t, results[0].LoginName, tt.expectedLoginName)
    })
}
```

Reply:
Ooh, nice. This will make it much easier to (later) expand the test to cover the shibboleth provider and group search behaviors.